src/sql/src/pure.rs

```diff
@@ -161,6 +161,22 @@ pub async fn purify_create_source(
         }
     };
 
+    match &connection {
+        CreateSourceConnection::Kafka(_)
+        | CreateSourceConnection::Kinesis { .. }
+        | CreateSourceConnection::S3 { .. }
+        | CreateSourceConnection::TestScript { .. } => match &referenced_subsources {
+            Some(ReferencedSubsources::All) => {
+                sql_bail!("FOR ALL TABLES is only valid for multi-output sources");
```

Comment:
This will cause existing envds with this to not boot. Is that ok?

Also: is it possible to enforce this directly in the type system in the AST node itself (i.e., in the sql-parser crate)? That would make it impossible for us to forget to check for it.

Reply:
@mjibson statements like this trigger and internal error, instead of a user-facing error (https://github.com/MaterializeInc/materialize/issues/15499#issuecomment-1283395708), this is just checking that error way earlier (as opposed to after purification)

We are avoiding touching the parser here because `FOR ALL TABLES` and `FOR TABLES` are used by multiple source types, and so are parsed at the end of the statement, for all sources. Fixing that is a much more intrusive change!